File: vx01_ws/src/vx01_control/locomotion/vx01_locomotion_control/src/teleop_node.cpp
#include "vx01_locomotion_control/teleop_node.hpp"

#include <unistd.h>
#include <cstdio>
#include <thread>
#include <iostream>

namespace vx01_locomotion_control {

static constexpr char KEY_W     = 'w';
static constexpr char KEY_S     = 's';
static constexpr char KEY_A     = 'a';
static constexpr char KEY_D     = 'd';
static constexpr char KEY_Q     = 'q';
static constexpr char KEY_E     = 'e';
static constexpr char KEY_SPACE = ' ';
static constexpr char KEY_X     = 'x';

TeleopNode::TeleopNode(const rclcpp::NodeOptions & options)
: Node("teleop_node", options)
{
    declare_parameter("linear_step",    0.02);
    declare_parameter("angular_step",   0.1);
    declare_parameter("max_linear_vel", 0.15);
    declare_parameter("max_angular_vel",0.5);

    linear_step_    = get_parameter("linear_step").as_double();
    angular_step_   = get_parameter("angular_step").as_double();
    max_linear_vel_ = get_parameter("max_linear_vel").as_double();
    max_angular_vel_= get_parameter("max_angular_vel").as_double();

    cmd_vel_pub_ = create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

    pub_timer_ = create_wall_timer(
        std::chrono::milliseconds(100),
        std::bind(&TeleopNode::publishVelocity, this));

    struct termios raw;
    tcgetattr(STDIN_FILENO, &orig_termios_);
    terminal_saved_ = true;
    raw = orig_termios_;
    raw.c_lflag &= ~(ICANON | ECHO);
    raw.c_cc[VMIN]  = 0;
    raw.c_cc[VTIME] = 1;
    tcsetattr(STDIN_FILENO, TCSANOW, &raw);

    printHelp();

    std::thread([this]() { readLoop(); }).detach();
}

TeleopNode::~TeleopNode()
{
    restoreTerminal();
}

void TeleopNode::restoreTerminal()
{
    if (terminal_saved_) {
        tcsetattr(STDIN_FILENO, TCSANOW, &orig_termios_);
        terminal_saved_ = false;
    }
}

char TeleopNode::getKey()
{
    char c = 0;
    read(STDIN_FILENO, &c, 1);
    return c;
}

void TeleopNode::readLoop()
{
    while (rclcpp::ok()) {
        char c = getKey();
        if (c == 0) continue;

        bool changed = true;
        switch (c) {
            case KEY_W:
                vx_ = std::clamp(vx_ + linear_step_, -max_linear_vel_, max_linear_vel_);
                break;
            case KEY_S:
                vx_ = std::clamp(vx_ - linear_step_, -max_linear_vel_, max_linear_vel_);
                break;
            case KEY_A:
                omega_ = std::clamp(omega_ + angular_step_, -max_angular_vel_, max_angular_vel_);
                break;
            case KEY_D:
                omega_ = std::clamp(omega_ - angular_step_, -max_angular_vel_, max_angular_vel_);
                break;
            case KEY_Q:
                vy_ = std::clamp(vy_ + linear_step_, -max_linear_vel_, max_linear_vel_);
                break;
            case KEY_E:
                vy_ = std::clamp(vy_ - linear_step_, -max_linear_vel_, max_linear_vel_);
                break;
            case KEY_SPACE:
                vx_ = vy_ = omega_ = 0.0;
                break;
            case KEY_X:
                vx_ = vy_ = omega_ = 0.0;
                rclcpp::shutdown();
                break;
            default:
                changed = false;
                break;
        }

        if (changed) {
            RCLCPP_INFO(get_logger(),
                "vx=%.3f  vy=%.3f  omega=%.3f", vx_, vy_, omega_);
        }
    }
}

void TeleopNode::publishVelocity()
{
    geometry_msgs::msg::Twist msg;
    msg.linear.x  = vx_;
    msg.linear.y  = vy_;
    msg.angular.z = omega_;
    cmd_vel_pub_->publish(msg);
}

void TeleopNode::printHelp() const
{
    std::cout << R"(
  ╔══════════════════════════════════════╗
  ║   VX01 Hexapod Teleop               ║
  ╠══════════════════════════════════════╣
  ║  W / S   — forward / backward       ║
  ║  A / D   — turn left / turn right   ║
  ║  Q / E   — strafe left / right      ║
  ║  SPACE   — stop (zero velocity)     ║
  ║  X       — stop and exit            ║
  ╚══════════════════════════════════════╝
)" << std::endl;
}

}  // namespace vx01_locomotion_control

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<vx01_locomotion_control::TeleopNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: vx01_ws/src/vx01_control/locomotion/vx01_hexapod_locomotion/vx01_hexapod_locomotion/teleop_node.py
import sys
import tty
import termios

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist


class TeleopNode(Node):

    LINEAR_STEP = 0.05
    ANGULAR_STEP = 0.3
    MAX_LINEAR = 0.20
    MAX_ANGULAR = 0.60

    def __init__(self):
        super().__init__('teleop_node')
        self._pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self._vx = 0.0
        self._vy = 0.0
        self._omega = 0.0

    def _get_key(self) -> str:
        fd = sys.stdin.fileno()
        old = termios.tcgetattr(fd)
        try:
            tty.setraw(fd)
            key = sys.stdin.read(1)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old)
        return key

    def _clamp(self, val: float, limit: float) -> float:
        return max(-limit, min(limit, val))

    def _publish(self):
        msg = Twist()
        msg.linear.x = self._vx
        msg.linear.y = self._vy
        msg.angular.z = self._omega
        self._pub.publish(msg)
        print(
            f'\r  vx={self._vx:+.2f}  vy={self._vy:+.2f}  omega={self._omega:+.2f}  ',
            end='',
            flush=True,
        )

    def _stop(self):
        self._vx = 0.0
        self._vy = 0.0
        self._omega = 0.0

    def run(self):
        print('VX-01 Hexapod Teleop')
        print('─' * 38)
        print('  W / S   →  forward / backward')
        print('  A / D   →  turn left / turn right')
        print('  Q / E   →  strafe left / strafe right')
        print('  SPACE   →  stop')
        print('  Ctrl+C  →  quit')
        print('─' * 38)
        self._publish()

        while True:
            key = self._get_key()

            if key == 'w':
                self._vx += self.LINEAR_STEP
            elif key == 's':
                self._vx -= self.LINEAR_STEP
            elif key == 'a':
                self._omega += self.ANGULAR_STEP
            elif key == 'd':
                self._omega -= self.ANGULAR_STEP
            elif key == 'q':
                self._vy += self.LINEAR_STEP
            elif key == 'e':
                self._vy -= self.LINEAR_STEP
            elif key == ' ':
                self._stop()
            elif key == '\x03':
                break

            self._vx = self._clamp(self._vx, self.MAX_LINEAR)
            self._vy = self._clamp(self._vy, self.MAX_LINEAR)
            self._omega = self._clamp(self._omega, self.MAX_ANGULAR)

            self._publish()


def main(args=None):
    rclpy.init(args=args)
    node = TeleopNode()
    try:
        node.run()
    except KeyboardInterrupt:
        pass
    node._stop()
    node._publish()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
